File: apps/mobile/android/app/src/main/kotlin/com/rarocamera/raro_mobile/camera/CameraErrorMapper.kt
package com.rarocamera.raro_mobile.camera

import com.rarocamera.raro_mobile.generated.camera.CameraErrorCode

sealed class CameraNativeException(message: String? = null) : Exception(message) {
  object PermissionDenied : CameraNativeException()
  object DeviceUnavailable : CameraNativeException()
  object LensUnavailable : CameraNativeException()
  object FormatUnsupported : CameraNativeException()
  class SessionFailed(msg: String) : CameraNativeException(msg)
  object AlreadyRunning : CameraNativeException()
  object NotRunning : CameraNativeException()
}

fun CameraNativeException.toCode(): CameraErrorCode = when (this) {
  is CameraNativeException.PermissionDenied -> CameraErrorCode.PERMISSION_DENIED
  is CameraNativeException.DeviceUnavailable -> CameraErrorCode.DEVICE_UNAVAILABLE
  is CameraNativeException.LensUnavailable -> CameraErrorCode.LENS_UNAVAILABLE
  is CameraNativeException.FormatUnsupported -> CameraErrorCode.FORMAT_UNSUPPORTED
  is CameraNativeException.SessionFailed -> CameraErrorCode.SESSION_FAILED
  is CameraNativeException.AlreadyRunning -> CameraErrorCode.ALREADY_RUNNING
  is CameraNativeException.NotRunning -> CameraErrorCode.NOT_RUNNING
}

fun CameraNativeException.symbolicCode(): String = when (this) {
  is CameraNativeException.PermissionDenied -> "permissionDenied"
  is CameraNativeException.DeviceUnavailable -> "deviceUnavailable"
  is CameraNativeException.LensUnavailable -> "lensUnavailable"
  is CameraNativeException.FormatUnsupported -> "formatUnsupported"
  is CameraNativeException.SessionFailed -> "sessionFailed"
  is CameraNativeException.AlreadyRunning -> "alreadyRunning"
  is CameraNativeException.NotRunning -> "notRunning"
}

File: apps/mobile/ios/Runner/Native/Camera/CameraErrorMapper.swift
import AVFoundation
import Foundation

enum CameraNativeError: Error {
  case permissionDenied
  case deviceUnavailable
  case lensUnavailable
  case formatUnsupported
  case sessionFailed(String)
  case alreadyRunning
  case notRunning
  case sessionInterrupted
}

extension CameraNativeError {
  var code: CameraErrorCode {
    switch self {
    case .permissionDenied: return .permissionDenied
    case .deviceUnavailable: return .deviceUnavailable
    case .lensUnavailable: return .lensUnavailable
    case .formatUnsupported: return .formatUnsupported
    case .sessionFailed: return .sessionFailed
    case .alreadyRunning: return .alreadyRunning
    case .notRunning: return .notRunning
    case .sessionInterrupted: return .sessionInterrupted
    }
  }

  var message: String? {
    if case let .sessionFailed(msg) = self { return msg }
    return nil
  }
}
